File: main/src/main/scala/org/clulab/dynet/TaskManager.scala
package org.clulab.dynet

import com.typesafe.config.Config
import org.clulab.dynet.TaskManager._
import org.clulab.sequences.{ColumnReader, Row}
import org.clulab.utils.{Configured, MathUtils}
import org.slf4j.{Logger, LoggerFactory}

import scala.collection.mutable.ArrayBuffer
import scala.util.Random

/**
  * Manages the tasks in LstmCrfMtl
  */
class TaskManager(config:Config) extends Configured {
  override def getConf: Config = config

  /** How many shards to have per epoch */
  val shardsPerEpoch:Int = getArgInt("mtl.shardsPerEpoch", Some(10))

  /** Total number of epochs */
  val maxEpochs:Int = getArgInt("mtl.maxEpochs", Some(100))

  /** Training patience in number of epochs */
  val epochPatience:Int = getArgInt("mtl.epochPatience", Some(5))

  /** Array of all tasks to be managed */
  val tasks:Array[Task] = readTasks()

  /** Training shards from all tasks */
  val shards:Array[Shard] = mkShards()

  def taskCount:Int = tasks.length
  def indices: Range = tasks.indices

  /** Construct training shards by interleaving shards from all tasks */
  private def mkShards():Array[Shard] = {
    val shardsByTasks = new Array[Array[Shard]](tasks.length)

    // construct the shards for each task
    for(i <- tasks.indices) {
      shardsByTasks(i) = tasks(i).mkShards()
      assert(shardsByTasks(i).length == shardsPerEpoch)
    }

    // now interleave the tasks
    val interleavedShards = new ArrayBuffer[Shard]()
    for(i <- 0 until shardsPerEpoch) {
      for(j <- tasks.indices) {
        val crtShard = shardsByTasks(j)(i)
        interleavedShards += crtShard
      }
    }

    /*
    logger.debug("All shards:")
    for(i <- interleavedShards.indices) {
      logger.debug(s"${interleavedShards(i)}")
    }
    */

    interleavedShards.toArray
  }

  /** Iterator over all sentences coming from all interleaved shards */
  def getSentences(random:Random):Iterator[(Int, Array[Row])] = new SentenceIterator(tasks, shards, random)

  /** Reads all tasks from disk in memory */
  protected def readTasks(): Array[Task] = {
    val numberOfTasks = getArgInt("mtl.numberOfTasks", None)
    val tasks = new ArrayBuffer[Task]()
    for(i <- 0 until numberOfTasks) {
      tasks += readTask(i + 1)
    }
    logger.debug(s"Read $numberOfTasks tasks from config file.")
    tasks.toArray
  }

  protected def readTask(taskNumber: Int): Task = {
    val taskName = getArgString(s"mtl.task$taskNumber.name", None)
    val train = getArgString(s"mtl.task$taskNumber.train", None)

    val dev =
      if(contains(s"mtl.task$taskNumber.dev")) Some(getArgString(s"mtl.task$taskNumber.dev", None))
      else None

    val test =
      if(contains(s"mtl.task$taskNumber.test")) Some(getArgString(s"mtl.task$taskNumber.test", None))
      else None

    val taskType = parseType(getArgString(s"mtl.task$taskNumber.type", Some("basic")))

    val weight:Float =
      getArgFloat(s"mtl.task$taskNumber.weight", Some(1.0f))

    new Task(taskNumber - 1, taskName, train, dev, test, taskType, shardsPerEpoch, weight)
  }

  def parseType(inf: String): Int = inf match {
    case "basic" => TaskManager.TYPE_BASIC
    case "dual" => TaskManager.TYPE_DUAL
    case _ => throw new RuntimeException(s"ERROR: unknown task type $inf!")
  }

  def debugTraversal():Unit = {
    val rand = new Random()
    for(epoch <- 0 until maxEpochs) {
      logger.debug(s"Started epoch $epoch")
      var sentCount = 0
      var taskId = 0
      var totalSents = 0
      for(sentence <- getSentences(rand)) {
        totalSents += 1
        if(sentence._1 != taskId) {
          logger.debug(s"Read $sentCount sentences from task $taskId")
          taskId = sentence._1
          sentCount = 1
        } else {
          sentCount += 1
        }
      }
      logger.debug(s"Read $sentCount sentences from task $taskId")
      logger.debug(s"Read $totalSents sentences in epoch $epoch.")
    }
  }
}

class SentenceIterator(val tasks:Array[Task], val shards:Array[Shard], val random:Random)
  extends Iterator[(Int, Array[Row])] {
  /** Offset in randomizedSentencePositions array */
  private var sentenceOffset:Int = 0

  private val randomizedSentencePositions:Array[Sentence] = randomizeSentences()

  private case class Sentence(taskId:Int, sentencePosition:Int)

  /** Randomizes all sentences across all tasks */
  private def randomizeSentences(): Array[Sentence] = {
    // first, randomize the shards
    val randomizedShards = MathUtils.randomize(shards, random)
    val randomizedSents = new ArrayBuffer[Sentence]()
    for(shard <- randomizedShards) {
      // second, randomize the sentences inside each shard
      val sents = MathUtils.randomize((shard.startPosition until shard.endPosition).toArray, random)
      for(sent <- sents) {
        // store the randomized sentences
        randomizedSents += Sentence(shard.taskId, sent)
      }
    }
    randomizedSents.toArray
  }

  override def hasNext: Boolean = sentenceOffset < randomizedSentencePositions.length

  override def next(): (Int, Array[Row]) = {
    assert(sentenceOffset >= 0 && sentenceOffset < randomizedSentencePositions.length)

    val s = randomizedSentencePositions(sentenceOffset)
    val tid = s.taskId
    val sentence = tasks(tid).trainSentences(s.sentencePosition)
    sentenceOffset += 1

    //logger.debug(s"shardPosition = $shardPosition, sentencePosition = $sentencePosition")

    (tid, sentence)
  }

}

case class Shard(taskId:Int, startPosition:Int, endPosition:Int)

class Task(
  val taskId:Int, // this starts at 0 so we can use it as an index in the array of tasks
  val taskName:String,
  val trainFileName:String,
  val devFileName:Option[String],
  val testFileName:Option[String],
  val taskType:Int,
  val shardsPerEpoch:Int,
  val taskWeight:Float) {
  logger.debug(s"Reading task $taskNumber ($taskName)...")
  val trainSentences:Array[Array[Row]] = ColumnReader.readColumns(trainFileName)
  val devSentences:Option[Array[Array[Row]]] =
    if(devFileName.isDefined) Some(ColumnReader.readColumns(devFileName.get))
    else None
  val testSentences:Option[Array[Array[Row]]] =
    if(testFileName.isDefined) Some(ColumnReader.readColumns(testFileName.get))
    else None

  val isBasic:Boolean = taskType == TaskManager.TYPE_BASIC
  val isDual:Boolean = taskType == TaskManager.TYPE_DUAL

  def prettyType: String =
    taskType match {
      case TaskManager.TYPE_BASIC => "basic"
      case TaskManager.TYPE_DUAL => "dual"
      case _ => "unknown"
    }

  // The size of the training shard for this task
  val shardSize:Int = math.ceil(trainSentences.length.toDouble / shardsPerEpoch).toInt

  // Current position in the training sentences when we iterate during training
  var currentTrainingSentencePosition:Int = 0

  logger.debug(s"============ starting task $taskNumber ============")
  logger.debug(s"Read ${trainSentences.length} training sentences for task $taskNumber, with shard size $shardSize.")
  if(devSentences.isDefined)
    logger.debug(s"Read ${devSentences.get.length} development sentences for task $taskNumber.")
  if(testSentences.isDefined)
    logger.debug(s"Read ${testSentences.get.length} testing sentences for task $taskNumber.")
  logger.debug(s"Using taskWeight = $taskWeight")
  logger.debug(s"Task type = $prettyType.")
  logger.debug(s"============ completed task $taskNumber ============")

  /** Construct the shards from all training sentences in this task */
  def mkShards():Array[Shard] = {
    val shards = new ArrayBuffer[Shard]()
    var crtPos = 0
    while(crtPos < trainSentences.length) {
      val endPos = math.min(crtPos + shardSize, trainSentences.length)
      shards += Shard(taskId, crtPos, endPos)
      crtPos = endPos
    }
    shards.toArray
  }

  def taskNumber:Int = taskId + 1
}

object TaskManager {
  val logger:Logger = LoggerFactory.getLogger(classOf[TaskManager])

  val TYPE_BASIC = 0
  val TYPE_DUAL = 1

}

File: main/src/main/python/pytorch/taskManager.py
import random
import math
from sequences.columnReader import ColumnReader
from dataclasses import dataclass

TYPE_BASIC = 0
TYPE_DUAL = 1

class TaskManager():

  def __init__(self, config, seed):

    self.config = config
    self.random = seed

    # How many shards to have per epoch
    self.shardsPerEpoch = config.get_int("mtl.shardsPerEpoch", 10)

    # Total number of epochs 
    self.maxEpochs:Int = config.get_int("mtl.maxEpochs", 100)

    # Training patience in number of epochs 
    self.epochPatience:Int = config.get_int("mtl.epochPatience", 5)

    # Array of all tasks to be managed 
    self.tasks = self.readTasks()

    self.taskCount = len(self.tasks)
    self.indices = range(self.taskCount)

    # Training shards from all tasks 
    self.shards = self.mkShards()

  def contains(self, paramPrefix):
    return self.config.__contains__(paramPrefix)

  def get_int(self, x, defualt=None):
    return self.config.get_int(x, defualt)

  def get_string(self, x, defualt=None):
    return self.config.get_string(x, defualt)

  def get_float(self, x, defualt=None):
    return self.config.get_float(x, defualt)

  def get_bool(self, x, defualt=None):
    return self.config.get_bool(x, defualt)

  def get_list(self, x, defualt=None):
    return self.config.get_list(x, defualt)

  def get_config(self, x, defualt=None):
    return self.config.get_config(x, defualt)

  # Construct training shards by interleaving shards from all tasks 
  def mkShards(self):
    shardsByTasks = list()

    # construct the shards for each task
    for i in self.indices:
      shardsByTasks += [self.tasks[i].mkShards()]
      assert(len(shardsByTasks[i]) == self.shardsPerEpoch)

    # now interleave the tasks
    interleavedShards = list()
    for i in range(self.shardsPerEpoch):
      for j in self.indices:
        crtShard = shardsByTasks[j][i]
        interleavedShards += [crtShard]

    return interleavedShards

  # Iterator over all sentences coming from all interleaved shards 
  def getSentences(self):
    random.seed(self.random)
    randomizedShards = random.sample(self.shards, len(self.shards))
    for shard in randomizedShards:
      sents = random.sample(range(shard.startPosition, shard.endPosition), shard.endPosition-shard.startPosition)
      for sent in sents:
        yield (shard.taskId, self.tasks[shard.taskId].trainSentences[sent])

  # Reads all tasks from disk in memory 
  def readTasks(self):
    numberOfTasks = self.config.get_int("mtl.numberOfTasks", None)
    tasks = list()
    for i in range(numberOfTasks):
      tasks += [self.readTask(i + 1)]

    print (f"Read {numberOfTasks} tasks from config file.")
    return tasks

  def readTask(self, taskNumber):
    taskName = self.config.get_string(f"mtl.task{taskNumber}.name", None)
    train = self.config.get_string(f"mtl.task{taskNumber}.train", None)

    dev = self.config.get_string(f"mtl.task{taskNumber}.dev", None) if f"mtl.task{taskNumber}.dev" in self.config else None
    test = self.config.get_string(f"mtl.task{taskNumber}.test", None) if f"mtl.task{taskNumber}.test" in self.config else None

    taskType = self.parseType(self.config.get_string(f"mtl.task{taskNumber}.type", "basic"))

    weight = self.config.get_float(f"mtl.task{taskNumber}.weight", 1.0)

    return Task(taskNumber - 1, taskName, taskType, self.shardsPerEpoch, weight, train, dev, test)

  def parseType(self, inf):
    if inf == "basic": return TYPE_BASIC
    elif inf == "dual": return TYPE_DUAL
    else: raise ValueError(f"ERROR: unknown task type {inf}!")

  def debugTraversal(self):
    for epoch in range(self.maxEpochs):
      print (f"Started epoch {epoch}")
      sentCount = 0
      taskId = 0
      totalSents = 0
      for sentence in self.getSentences():
        totalSents += 1
        if(sentence[0] != taskId):
          print (f"Read {sentCount} sentences from task {taskId}")
          taskId = sentence[0]
          sentCount = 1
        else:
          sentCount += 1
      print (f"Read {sentCount} sentences from task {taskId}")
      print (f"Read {totalSents} sentences in epoch {epoch}.")

@dataclass
class Shard:
  taskId: int
  startPosition: int
  endPosition: int

class Task:
  def __init__(self,
  taskId, # this starts at 0 so we can use it as an index in the array of tasks
  taskName:str,
  taskType:int,
  shardsPerEpoch:int,
  taskWeight:float,
  trainFileName:str,
  devFileName:str = None,
  testFileName:str = None):
    self.taskId = taskId
    taskNumber = taskId + 1
    print (f"Reading task {taskNumber} ({taskName})...")
    self.trainSentences = ColumnReader.readColumns(trainFileName)
    l = len(self.trainSentences)
    self.trainSentences = self.trainSentences
    self.devSentences = ColumnReader.readColumns(devFileName) if devFileName else None
    self.testSentences = ColumnReader.readColumns(testFileName) if testFileName else None

    self.isBasic:bool = taskType == TYPE_BASIC
    self.isDual:bool = taskType == TYPE_DUAL

    if taskType == TYPE_BASIC: 
      self.prettyType = "basic"
    elif taskType == TYPE_DUAL: 
      self.prettyType = "dual"
    else: 
      self.prettyType = "unknown"

    # The size of the training shard for this task
    self.shardSize = math.ceil(len(self.trainSentences) / shardsPerEpoch)

    # Current position in the training sentences when we iterate during training
    currentTrainingSentencePosition = 0

    self.taskWeight = taskWeight
    self.taskName = taskName

    print (f"============ starting task {taskNumber} ============")
    print (f"Read {len(self.trainSentences)} training sentences for task {taskNumber}, with shard size {self.shardSize}.")
    if(self.devSentences is not None):
      print (f"Read {len(self.devSentences)} development sentences for task {taskNumber}.")
    if(self.testSentences is not None):
      print (f"Read {len(self.testSentences)} testing sentences for task {taskNumber}.")
    print (f"Using taskWeight = {taskWeight}")
    print (f"Task type = {self.prettyType}.")
    print (f"============ completed task {taskNumber} ============")

  # Construct the shards from all training sentences in this task 
  def mkShards(self):
    shards = list()
    crtPos = 0
    while(crtPos < len(self.trainSentences)):
      endPos = min(crtPos + self.shardSize, len(self.trainSentences))
      shards += [Shard(self.taskId, crtPos, endPos)]
      crtPos = endPos
    return shards
